File: pages/refactorings/extract-method/03-expression-eval/After.scala
// Evaluate an arithmetic expression; division by zero has no value.
object After:
  enum Expr:
    case Lit(n: Int)
    case Add(l: Expr, r: Expr)
    case Mul(l: Expr, r: Expr)
    case Div(l: Expr, r: Expr)
  import Expr.*

  def eval(e: Expr): Option[Int] = e match
    case Lit(n)    => Some(n)
    case Add(l, r) => binary(l, r)((a, b) => Some(a + b))
    case Mul(l, r) => binary(l, r)((a, b) => Some(a * b))
    case Div(l, r) => binary(l, r)((a, b) => Option.when(b != 0)(a / b))

  // Takes the operands unevaluated, so the right one is still only evaluated
  // when the left one has a value, exactly as in Before.
  def binary(l: Expr, r: Expr)(op: (Int, Int) => Option[Int]): Option[Int] =
    for a <- eval(l); b <- eval(r); c <- op(a, b) yield c

File: pages/refactorings/extract-method/03-expression-eval/After.hs
-- Evaluate an arithmetic expression; division by zero has no value.
module After where

data Expr = Lit Int | Add Expr Expr | Mul Expr Expr | Div Expr Expr
  deriving Show

eval :: Expr -> Maybe Int
eval (Lit n)   = Just n
eval (Add l r) = binary (\a b -> Just (a + b)) l r
eval (Mul l r) = binary (\a b -> Just (a * b)) l r
eval (Div l r) = binary (\a b -> if b == 0 then Nothing else Just (a `div` b)) l r

-- Takes the operands unevaluated, so the right one is still only forced
-- when the left one has a value, exactly as in Before.
binary :: (Int -> Int -> Maybe Int) -> Expr -> Expr -> Maybe Int
binary op l r = do
  a <- eval l
  b <- eval r
  op a b
